### PR TITLE
Adds create user API with email validation.

### PR DESCRIPTION
**Pull request by gaurav-jondhale**

### CODE REVIEW

File: repository/user_repository.go [hunk 1]
```diff
@@ -46,9 +52,23 @@ func (repo *userRepo) FindUser(ctx context.Context, email string) (*domain.User,
...
 
 func (repo *userRepo) CreateUser(ctx context.Context, user domain.User) (*domain.User, error) {
-	return nil, nil
-}
+	var user1 domain.User
+	email := user.Email
+
+	valid := emailRegex.MatchString(email)
+
+	if valid == false {
```
Comment (vkhichar) [inline repository/user_repository.go]: request validation should be at handler layer

File: contract/create_user_request.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+	valid := emailRegex.MatchString(email)
+
+	if valid == false {
+
+		return errors.New("invalid email")
+
+	}
+
+	if strings.TrimSpace(req.Email) == "" {
```
Comment (vkhichar) [inline contract/create_user_request.go]: will this code be reachable in any case?

File: handler/routes.go [hunk 3]
```diff
@@ -9,6 +9,7 @@ func Routes() *mux.Router {
 
 	router.HandleFunc("/ping", PingHandler())
 	router.HandleFunc("/login", LoginHandler(deps.userService)).Methods("POST")
+	router.Handle("/create_user", CreateUserHandler(deps.userService))
```
Comment (vkhichar) [inline handler/routes.go]: route should be `POST /users`

File: handler/user_handler.go [hunk 4]
```diff
@@ -60,3 +61,68 @@ func LoginHandler(userService service.UserService) http.HandlerFunc {
...
+		if err != nil {
+			fmt.Printf("handler: invalid request for email: %s", req.Email)
+
+			w.WriteHeader(http.StatusBadRequest)
+			responseBytes, _ := json.Marshal(contract.ErrorResponse{Error: err.Error()})
+			w.Write(responseBytes)
+			return
+		}
+
+		user1 := domain.User{
```
Comment (vkhichar) [inline handler/user_handler.go]: use better/readable variable names

may be use: user, createdUser

File: repository/db.go [hunk 5]
```diff
@@ -11,6 +11,7 @@ import (
 var sqlxDB *sqlx.DB
 
 func InitDB() {
+	fmt.Println(getConnectionString())
```
Comment (vkhichar) [inline repository/db.go]: Remove un-necessary code

File: repository/user_repository.go [hunk 6]
```diff
@@ -46,9 +46,15 @@ func (repo *userRepo) FindUser(ctx context.Context, email string) (*domain.User,
 }
 
 func (repo *userRepo) CreateUser(ctx context.Context, user domain.User) (*domain.User, error) {
-	return nil, nil
-}
+	var user1 domain.User
```
Comment (vkhichar) [inline repository/user_repository.go]: use better variable names

File: contract/create_user_request.go (hunk 2 above)
Comment (gaurav-jondhale) [reply contract/create_user_request.go]: yes I think I tested it.

Comment (vkhichar) [reply contract/create_user_request.go]: if the email is empty then that will be returned from above line where we are validating the email by regular expression

File: handler/routes.go [hunk 7]
```diff
@@ -9,6 +9,7 @@ func Routes() *mux.Router {
 
 	router.HandleFunc("/ping", PingHandler())
 	router.HandleFunc("/login", LoginHandler(deps.userService)).Methods("POST")
+	router.Handle("/create_user", CreateUserHandler(deps.userService)).Methods("POST")
```
Comment (wsushmita) [inline handler/routes.go]: change route to /users

File: handler/user_handler.go [hunk 8]
```diff
@@ -60,3 +61,68 @@ func LoginHandler(userService service.UserService) http.HandlerFunc {
...
+		var req contract.CreateUserRequest
+
+		err := json.NewDecoder(r.Body).Decode(&req)
+
+		if err != nil {
+
+			fmt.Printf("handler: error while decoding request for create user: %s", err.Error())
+
+			w.WriteHeader(http.StatusBadRequest)
+			responseBytes, _ := json.Marshal(contract.ErrorResponse{Error: "invalid request"})
```
Comment (wsushmita) [inline handler/user_handler.go]: don't ignore err with _

File: handler/user_handler.go [hunk 9]
```diff
@@ -60,3 +61,68 @@ func LoginHandler(userService service.UserService) http.HandlerFunc {
...
+			w.Write(responseBytes)
+			return
+		}
+
+		err = req.Validate()
+		if err != nil {
+			fmt.Printf("handler: invalid request for email: %s", req.Email)
+
+			w.WriteHeader(http.StatusBadRequest)
+			responseBytes, _ := json.Marshal(contract.ErrorResponse{Error: err.Error()})
```
Comment (wsushmita) [inline handler/user_handler.go]: don't ignore error

File: handler/user_handler.go [hunk 10]
```diff
@@ -60,3 +61,68 @@ func LoginHandler(userService service.UserService) http.HandlerFunc {
...
+			IsAdmin:  req.IsAdmin,
+		}
+
+		user, err := userService.CreateUser(r.Context(), createdUser)
+
+		if err == service.ErrInvalidEmailPassword {
+			fmt.Printf("handler: invalid email or password for email: %s", req.Email)
+
+			w.WriteHeader(http.StatusUnauthorized)
+			responseBytes, _ := json.Marshal(contract.ErrorResponse{Error: "invalid email or password"})
```
Comment (wsushmita) [inline handler/user_handler.go]: don't ignore error

File: handler/user_handler.go [hunk 11]
```diff
@@ -60,3 +61,68 @@ func LoginHandler(userService service.UserService) http.HandlerFunc {
...
+			w.Write(responseBytes)
+			return
+		}
+
+		if err != nil {
+			fmt.Printf("handler: error while creating user for email: %s, error: %s", req.Email, err.Error())
+
+			w.WriteHeader(http.StatusInternalServerError)
+			responseBytes, _ := json.Marshal(contract.ErrorResponse{Error: "handler:something went wrong"})
```
Comment (wsushmita) [inline handler/user_handler.go]: don't ignore error

File: handler/user_handler.go [hunk 12]
```diff
@@ -60,3 +61,68 @@ func LoginHandler(userService service.UserService) http.HandlerFunc {
...
+			w.Write(responseBytes)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+		responseBytes, _ := json.Marshal(contract.CreateUserResponse{Name: user.Name, Email: user.Email, Password: user.Password, IsAdmin: user.IsAdmin})
```
Comment (wsushmita) [inline handler/user_handler.go]: don't ignore error

Comment (wsushmita) [inline handler/user_handler.go]: u can send directly user to json.marshal

File: repository/user_repository.go [hunk 13]
```diff
@@ -11,12 +11,12 @@ import (
 
 const (
 	getUserByEmailQuery = "SELECT id, name, email, password, is_admin FROM users WHERE email= $1"
+	createUserByQuery   = "INSERT INTO users (name, email, password,is_admin) VALUES ($1, $2, $3, $4) RETURNING id, name, email, password, is_admin"
```
Comment (wsushmita) [inline repository/user_repository.go]: add consistent spacing after , add created_at & updated_at fields in return

**Review by wsushmita — CHANGES_REQUESTED** [review]

File: handler/routes.go [hunk 14]
```diff
@@ -8,7 +8,12 @@ func Routes() *mux.Router {
 	router := mux.NewRouter()
 	router.HandleFunc("/ping", PingHandler())
 	router.HandleFunc("/login", LoginHandler(deps.userService)).Methods("POST")
+
 	router.HandleFunc("/users", ListUsersHandler(deps.userService)).Methods("GET")
+
+	router.HandleFunc("/adduser", CreateUserHandler(deps.userService)).Methods("POST")
```
Comment (vkhichar) [inline handler/routes.go]: route should be `POST /users`

**Review by vkhichar — CHANGES_REQUESTED** [review]

**Pull request merged**